presto-main-base/src/main/java/com/facebook/presto/sql/relational/FunctionResolution.java

```diff
@@ -302,7 +302,10 @@ public boolean isTryFunction(FunctionHandle functionHandle)
 
     public boolean isFailFunction(FunctionHandle functionHandle)
     {
-        return functionAndTypeResolver.getFunctionMetadata(functionHandle).getName().equals(functionAndTypeResolver.qualifyObjectName(QualifiedName.of("fail")));
+        // todo: Revert this hack once constant folding support lands in C++.
+        // For now, we always look use the presto.default.fail function even when the default namespace is switched.
+        // This is done for consistency since the BuiltInNamespaceRewriter rewrites the functionHandles to presto.default functionHandles.
+        return functionAndTypeResolver.getFunctionMetadata(functionHandle).getName().equals(QualifiedObjectName.valueOf(JAVA_BUILTIN_NAMESPACE, "fail"));
```

Comment:
Maybe this doesn't make sense but how about making it 
`functionAndTypeResolver.getFunctionMetadata(functionHandle).getName().equals(functionAndTypeResolver.qualifyObjectName(QualifiedName.of("fail"))) || functionAndTypeResolver.getFunctionMetadata(functionHandle).getName().equals(QualifiedObjectName.valueOf(JAVA_BUILTIN_NAMESPACE, "fail"))`

Ideally using
` isNativeExecution && functionAndTypeResolver.getFunctionMetadata(functionHandle).getName().equals(QualifiedObjectName.valueOf(JAVA_BUILTIN_NAMESPACE, "fail")) `would be more accurate